Wrappers/Python/ccpi/optimisation/algorithms/CGLS.py

```diff
@@ -34,98 +34,145 @@ class CGLS(Algorithm):
       x_init: initial guess
       operator: operator for forward/backward projections
       data: data to operate on
+      tolerance: tolerance to stop algorithm
+      
+    Reference:
+        https://web.stanford.edu/group/SOL/software/cgls/
+      
     '''
+    
+    
+    
     def __init__(self, **kwargs):
+        
         super(CGLS, self).__init__()
         self.x        = kwargs.get('x_init', None)
         self.operator = kwargs.get('operator', None)
         self.data     = kwargs.get('data', None)
+        self.tolerance     = kwargs.get('tolerance', None)
         if self.x is not None and self.operator is not None and \
            self.data is not None:
-            print ("Calling from creator")
             self.set_up(x_init  =kwargs['x_init'],
                                operator=kwargs['operator'],
                                data    =kwargs['data'])
-
+            
+        if self.tolerance is None:
+            self.tolerance = 1e-6            
+                                    
     def set_up(self, x_init, operator , data ):
 
-        self.r = data.copy()
-        self.x = x_init * 0
-
-        self.operator = operator
-        self.d = operator.adjoint(self.r)
-
+        self.x = x_init
+        self.r = data - self.operator.direct(self.x)
+        self.s = self.operator.adjoint(self.r)
         
-        self.normr2 = self.d.squared_norm()
+        self.p = self.s
+        self.norms0 = self.s.norm()
+        
+        ##
+        self.norms = self.s.norm()
+        ##
+        
+        
+        self.gamma = self.norms0**2
```

Comment:
Why not use `squared_norm()`?

Reply:
`self.norms0` is calculated in the `set_up` and reused in the calculation of the `flag` to stop iterating. It makes sense to calculate the `norm` rather than `squared_norm` as it will be used more time as norm than squared.
What I find strange is that it is a constant?